src/metadata/metadata_io.c

```diff
@@ -218,106 +199,128 @@ static void metadata_io_req_drain(struct metadata_io_request *meta_io_req)
-			PAGES_TO_BYTES(meta_io_req->page),
-			PAGES_TO_BYTES(meta_io_req->count),
-			OCF_WRITE, 0, 0);
+			PAGES_TO_BYTES(m_req->page),
+			PAGES_TO_BYTES(m_req->count),
+			m_req->req.rw, 0, 0);
 	if (!io) {
-		metadata_io_i_asynch_end(meta_io_req, -OCF_ERR_NO_MEM);
+		metadata_io_io_end(m_req, -OCF_ERR_NO_MEM);
 		return 0;
 	}
 
 	/* Setup IO */
-	ocf_io_set_cmpl(io, meta_io_req, NULL, metadata_io_i_asynch_cmpl);
-	ret = ocf_io_set_data(io, meta_io_req->data, 0);
+	ocf_io_set_cmpl(io, m_req, NULL, metadata_io_io_cmpl);
+	ret = ocf_io_set_data(io, m_req->data, 0);
 	if (ret) {
 		ocf_io_put(io);
-		metadata_io_i_asynch_end(meta_io_req, ret);
+		metadata_io_io_end(m_req, ret);
 		return ret;
 	}
 	ocf_volume_submit_io(io);
 	return 0;
 }
 
+static struct ocf_io_if metadata_io_restart_if = {
+	.read = metadata_io_restart_req,
+	.write = metadata_io_restart_req,
+};
+
 /*
  * Iterative asynchronous write callback
  */
-static void metadata_io_i_asynch_end(struct metadata_io_request *request,
-		int error)
+static void metadata_io_io_end(struct metadata_io_request *m_req, int error)
 {
-	struct metadata_io_request_asynch *a_req;
-	ocf_cache_t cache;
-
-	OCF_CHECK_NULL(request);
-
-	cache = request->cache;
+	struct metadata_io_request_asynch *a_req = m_req->asynch;
+	ocf_cache_t cache = m_req->cache;
 
-	a_req = request->asynch;
 	OCF_CHECK_NULL(a_req);
 	OCF_CHECK_NULL(a_req->on_complete);
 
 	if (error) {
-		request->error |= error;
-		request->asynch->error |= error;
+		a_req->error = a_req->error ?: error;
 	} else {
-		if (request->fl_req.rw == OCF_READ)
-			metadata_io_req_drain(request);
+		if (m_req->req.rw == OCF_READ)
+			metadata_io_req_drain(m_req);
 	}
 
-	if (env_atomic_dec_return(&request->req_remaining))
-		return;
-
-	OCF_DEBUG_PARAM(cache, "Page = %u", request->page);
+	OCF_DEBUG_PARAM(cache, "Page = %u", m_req->page);
 
-	ctx_data_free(cache->owner, request->data);
-	request->data = NULL;
+	env_atomic_set(&m_req->finished, 1);
+	ocf_metadata_updater_kick(cache);
+}
 
-	if (env_atomic_dec_return(&a_req->req_remaining)) {
-		env_atomic_set(&request->finished, 1);
-		ocf_metadata_updater_kick(cache);
-		return;
-	}
+static uint32_t metadata_io_max_page(ocf_cache_t cache)
+{
+	return ocf_volume_get_max_io_size(&cache->device->volume) / PAGE_SIZE;
+}
 
-	OCF_DEBUG_MSG(cache, "Asynchronous IO completed");
+void metadata_io_req_end(struct metadata_io_request *m_req)
+{
+	struct metadata_io_request_asynch *a_req = m_req->asynch;
+	ocf_cache_t cache = m_req->cache;
 
-	/* All IOs have been finished, call IO end callback */
-	a_req->on_complete(request->cache, a_req->context, request->error);
+	ctx_data_free(cache->owner, m_req->data);
 
-	/*
-	 * If it's last request, we mark is as finished
-	 * after calling IO end callback
-	 */
-	env_atomic_set(&request->finished, 1);
-	ocf_metadata_updater_kick(cache);
+	if (env_atomic_dec_return(&a_req->req_remaining) == 0) {
+		a_req->on_complete(cache, a_req->context, a_req->error);
```

Comment:
It should be possible to move a_req->on_complete call to metadata_io_io_end. This way request will complete sooner. Deallocation would still be done here in metadata_io_req_end after request is removed from updater in_progress list.

Reply:
Good point. Done.